press/api/server.py

```diff
@@ -806,3 +806,37 @@ def create_record(action: str, scheduled_time: datetime) -> None:
 
 	create_record("Scale Up", formatted_scheduled_scale_up_time)
 	create_record("Scale Down", formatted_scheduled_scale_down_time)
+
+
+@frappe.whitelist()
+@protected(["Server", "Database Server"])
+def schedule_disk_resize(
+	name: str, server_type: str, volume_id: str, expected_volume_size: int | str, scheduled_datetime: str
+) -> None:
+	server = frappe.db.get_value(server_type, name, ["virtual_machine", "status"], as_dict=True)
+	if server.status != "Active":
+		frappe.throw(f"Cannot schedule disk resize for a server in {server.status} state")
+
+	if frappe.db.get_value(
+		"Virtual Disk Resize",
+		{
+			"creation": [">", frappe.utils.add_days(frappe.utils.now_datetime(), -7)],
+			"virtual_machine": server.virtual_machine,
+			"status": ["not in", ["Failure", "Cancelled"]],
+		},
+		as_dict=True,
+	):
+		frappe.throw(
+			"A disk resize was triggered within the last 7 days. Please wait before scheduling another resize."
+		)
+
+	frappe.get_doc(
+		{
+			"doctype": "Virtual Disk Resize",
+			"virtual_machine": server.virtual_machine,
+			"old_volume_id": volume_id,
+			"expected_volume_size": expected_volume_size,
+			"scheduled_datetime": scheduled_datetime,
+			"status": "Scheduled",
+		}
+	).insert()
```

Comment:
You can move this method on `BaseServer` class in `doctype/server/server.py` and apply the `@dashboard_whitelist()` decorator for permission checks.

Or, you can override the query condition and has_permission hook of `Virtual Disk Resize` as well
We have couple of those of other doctypes here : https://github.com/phot0n/press/blob/afec0fb33e65d4c6e47e3a394b869307684934df/press/hooks.py#L105-L158